src/Core/PostgreSQLProtocol.h

```diff
@@ -920,5 +1058,82 @@ class AuthenticationManager
 };
 }
 
+namespace PostgresPreparedStatements
+{
+
+class PreparedStatemetsManager
+{
+public:
+    explicit PreparedStatemetsManager(std::optional<size_t> limit_statements_)
+        : limit_statements(limit_statements_)
+    {
+    }
+
+    void addStatement(ASTPreparedStatement * statement)
+    {
+        if (limit_statements && statements.size() + 1 >= limit_statements.value())
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Statements limit exceeded");
+
+        statements[statement->function_name] = statement->function_body;
+    }
+
+    String getStatement(ASTExecute * execute)
+    {
+        return getStatement(execute->function_name, execute->arguments);
+    }
+
+    void deleteStatement(ASTDeallocate * query)
+    {
+        auto it = statements.find(query->function_name);
+        if (it == statements.end())
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Unknown statement");
+
+        statements.erase(it);
+    }
+
+    void attachBindQuery(std::unique_ptr<PostgreSQLProtocol::Messaging::BindQuery> query)
+    {
+        if (bind_query)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Query is already binded");
+
+        bind_query = std::move(query);
+    }
+
+    String getStatmentFromBind()
+    {
+        auto result = getStatement(bind_query->function_name, bind_query->parameters);
+
+        bind_query.reset();
+        return result;
+    }
+
+private:
+    std::unordered_map<String, String> statements;
+    std::optional<size_t> limit_statements;
+    std::unique_ptr<PostgreSQLProtocol::Messaging::BindQuery> bind_query;
```

Comment:
I'm not sure it's correct - if protocol supports multiple bindings for single execute then `bind_query` also should be a map - same as `statements`

Reply:
I think this is correct, because if you look at the messages from the postgres client under strace, the Parse, Bind, Describe, Execute requests will go together in one message. And since we use exactly one manager per channel, there will be no situation when 2 binds arrive

Reply:
Moreover, you can’t just use the map because execute query doesn’t know the name of the function